scalyr_agent/tests/monitors_manager_test.py

```diff
@@ -74,6 +83,133 @@ def test_module_with_id(self):
         self.assertEquals(test_manager.monitors[0].monitor_name, 'test_monitor(first)')
         self.assertEquals(test_manager.monitors[1].monitor_name, 'test_monitor(2)')
 
+    def test_user_agent_polling(self):
+        """Test polling of user-agent fragments and invocation of the callback (only on change)
+
+        A MonitorsManager + 2 monitors are started (each has their own threads).
+        1 monitor returns non-empty frags. 1 returns None.
+        We then verify that the callback is invoked correctly, only if the fragments change.
+
+        Notes:
+        - FakeClocks are used to avoid having to wait for thread loops.
+        - A FakeAgentLogger is defined so we can modify local test state
+        - Even though this test could have been broken up into 2 smaller ones, it is kept as one to minimize overhead
+            time taken by stop_manager()
+        """
+        counter = Counter()
+        counter_lock = threading.Lock()
+        test_frag = 'some_frag'
+
+        # Create MonitorsManager + 2 monitors. Set each to daemon otherwise unit test doesn't terminate
+        test_manager = self.__create_test_instance([
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+            {
+                'module': 'scalyr_agent.builtin_monitors.test_monitor',
+                'gauss_mean': 0,
+            },
+        ], [])
+        self.assertEquals(len(test_manager.monitors), 3)
+        patched_monitor_0 = test_manager.monitors[0]
+        patched_monitor_1 = test_manager.monitors[1]
+        unpatched_monitor = test_manager.monitors[2]
+
+        # Mock the function that returns user_agent_fragment (normally invoked on a Monitor)
+        def mock_get_user_agent_fragment():
+            # 2 patched monitors will call this function concurrently
+            counter_lock.acquire()
+            try:
+                counter['fragment_polls'] += 1
+            finally:
+                counter_lock.release();
+            return test_frag
+        for mon in [patched_monitor_0, patched_monitor_1]:
+            mon.get_user_agent_fragment = mock_get_user_agent_fragment
+            self.assertEquals(mon.get_user_agent_fragment(), test_frag)  # monkey patched
+        self.assertEquals(unpatched_monitor.get_user_agent_fragment(), None)  # not patched
+
+        # Mock the callback (that would normally be invoked on ScalyrClientSession)
+        # Check that the exact fragment returned by the 2 patched monitors are deduplicated.
+        def augment_user_agent(fragments):
+            counter['callback_invocations'] += 1
```

Comment:
Don't you have to hold the `counter_lock` here?

Reply:
I think we're ok here as only 1 thread accesses this state (`MonitorsManager`)

Reply:
I've moved the state that's accessed by multiple threads into the new `FragmentPolls` class. Hopefully this is easier to read/maintain.